### PR TITLE
Fix bad htlc add during reconnect

### PR DESCRIPTION
**Pull request by rustyrussell**
As found by random breakage on Travis on #3221 !

### CODE REVIEW

File: tests/test_connection.py [hunk 1]
```diff
@@ -2196,3 +2196,32 @@ def test_feerate_stress(node_factory, executor):
...
+                                                                  '-WIRE_COMMITMENT_SIGNED']}])
+
+        scid12 = l1.get_channel_scid(l2)
+        routel2l1 = [{'msatoshi': '10000msat', 'id': l1.info['id'], 'delay': 5, 'channel': scid12}]
+
+        # Get invoice from l1 to pay.
+        payhash1 = l1.rpc.invoice(10000, "invoice", "invoice")['payment_hash']
+    
+        # Start balancing payment.
+        fut = executor.submit(l1.pay, l2, 10**9 // 2)
```
Comment (niftynei) [inline tests/test_connection.py]: `check-python` is complaining about `fut` being unused here

File: tests/test_connection.py [hunk 2]
```diff
@@ -2196,3 +2196,32 @@ def test_feerate_stress(node_factory, executor):
...
+                                                  {'may_reconnect': True,
+                                                   'disconnect': ['=WIRE_UPDATE_ADD_HTLC',
+                                                                  '-WIRE_COMMITMENT_SIGNED']}])
+
+        scid12 = l1.get_channel_scid(l2)
+        routel2l1 = [{'msatoshi': '10000msat', 'id': l1.info['id'], 'delay': 5, 'channel': scid12}]
+
+        # Get invoice from l1 to pay.
+        payhash1 = l1.rpc.invoice(10000, "invoice", "invoice")['payment_hash']
+
```
Comment (niftynei) [inline tests/test_connection.py]: `check-python` is complaining about there being whitespace in this line

File: channeld/full_channel.c [hunk 3]
```diff
@@ -1316,16 +1351,21 @@ bool channel_force_htlcs(struct channel *channel,
 			htlc->failed_scid = NULL;
 	}
 
-	/* Now adjust balances.  The balance never goes negative, because
-	 * we do them in id order. */
+	/* Add giant offset so we never go negative here. */
+	if (!offset_balances(channel))
```
Comment (niftynei) [inline channeld/full_channel.c]: i assume we check an HTLC's validity (in terms of amount) before this point?

Comment (rustyrussell) [reply channeld/full_channel.c]: This code is only used when we're restoring HTLCs lightningd told us about.  So presumably they've been fully checked.

**Comment by rustyrussell** [comment]
To be clear: htlc restore amount checking is simply a sanity check.

To do the sanity check properly, we'd have to replay each HTLC state *in order*: once one has been fulfilled, for example, another can be spent.  It's not even enough to add them in creation order, we need every balance adjustment in order.  We don't have enough information to do that.

**Comment by rustyrussell** [comment]
flake8 now happy, trivial rebase.

**Comment by cdecker** [comment]
ACK https://github.com/ElementsProject/lightning/pull/3233/commits/eb012c036e144ac744ae0c501087aac15c09d266

**Pull request merged**